### DISCUSSION TITLE
how to use multi thread in filament

### DISCUSSION THREAD

**Discussion by Tony-qty** [google/filament#4875] [Q&A]
I am a iOS developer, I try to dispatch filament render in a new thread, but, there is a crash when I load my resource.
I use a pixelbuffer as nativewindow and set the flag with SWAP_CHAIN_CONFIG_APPLE_CVPIXELBUFFER. 

![image](https://user-images.githubusercontent.com/7675144/142582566-6a43eafa-7bfc-4133-8010-ce8218cf78b9.png)

**Comment by Tony-qty** [comment]
@bejado

**Reply by romainguy** [reply]
You must call Filament from a single thread. Filament itself is multithreaded though.

**Reply by Tony-qty** [reply]
> You must call Filament from a single thread. Filament itself is multithreaded though.

Can I dispatch filament  in a single thread queue.

**Reply by Tony-qty** [reply]
I create a thread queue, and dispatch all filament function in my queue. but , its till unable to run.

![image](https://user-images.githubusercontent.com/7675144/142601525-796e3342-66a5-4945-8fa3-0fe866e91c96.png)

**Reply by Tony-qty** [reply]
@romainguy

**Reply by bejado** [reply]
Somehow Filament is getting invoked from one or more different threads. All Filament APIs, including initialization, must be called from the same thread. You can use `std::this_thread::get_id()` to troubleshoot.
